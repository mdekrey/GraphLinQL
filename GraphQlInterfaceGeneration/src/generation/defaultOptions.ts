import { Options } from "./Options";

export const defaultOptions: Options = {
  namespace: "GraphLinqQL.Interfaces",
  useNullabilityIndicator: true,
  scalarTypes: {
    Int: { csharpNullable: false, csharpType: "int" },
    Float: { csharpNullable: false, csharpType: "double" },
    String: { csharpNullable: true, csharpType: "string" },
    Boolean: { csharpNullable: false, csharpType: "bool" }
  },
  using: [
    "GraphLinqQL",
    "GraphLinqQL.Introspection",
    "System",
    "System.Collections",
    "System.Collections.Generic",
    "System.Collections.Immutable",
    "System.Linq"
  ],
  deserializer: "System.Text.Json.JsonSerializer.Deserialize",
  introspection: false
};
