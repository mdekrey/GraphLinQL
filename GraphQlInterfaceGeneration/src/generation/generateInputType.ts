import { GraphQLInputObjectType, GraphQLInputField } from "graphql";
import { Options } from "./Options";
import { getTypeName } from "./getTypeName";
import { getPropertyName } from "./getPropertyName";
import { getInputTypeName } from "./getInputTypeName";
import { toCsharpValue } from "./toCsharpValue";

export function generateInputType(object: GraphQLInputObjectType, options: Options) {
  const typeName = getTypeName(object.name, options);

  const fields = object.getFields();
  return `${
    object.description
      ? `
/// <summary>
/// ${object.description.split("\n").join("\n/// ")}
/// </summary>`
      : `
`
  }
public class ${typeName}
{
    ${Object.keys(fields)
      .map(fieldName => fields[fieldName])
      .map(field => propertyDeclaration(field, options)).join(`
    `)}
}
`;
}

function propertyDeclaration(field: GraphQLInputField, options: Options) {
  const propertyName = getPropertyName(field.name, options);
  const typeName = getInputTypeName(field.type, options);
  const defaultValuePart = field.defaultValue ? ` = ${toCsharpValue(field.defaultValue, field.type, options)};` : "";
  return `${
    field.description
      ? `
      /// <summary>
      /// ${field.description.split("\n").join("\n    /// ")}
      /// </summary>
      `
      : ``
  }public ${typeName || "object"} ${propertyName} { get; set; }${defaultValuePart}`;
}
