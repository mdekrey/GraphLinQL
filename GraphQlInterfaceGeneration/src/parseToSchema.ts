import { parse, buildASTSchema, GraphQLSchema } from "graphql";

export function parseToSchema(graphQlSchema: string): GraphQLSchema {
  const astDocument = parse(graphQlSchema);

  const backcompatOptions = { commentDescriptions: true };
  const schema = buildASTSchema(astDocument, backcompatOptions);

  return schema;
}
