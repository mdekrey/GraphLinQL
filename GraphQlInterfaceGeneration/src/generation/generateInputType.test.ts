import { parseToSchema } from "../parseToSchema";
import { getStarWarsGraphQlSchemaV3 } from "../resources.test";
import { GraphQLInputObjectType } from "graphql";
import { defaultOptions } from "./defaultOptions";
import { generateInputType } from "./generateInputType";

it("can generate input types from v3", async function() {
  const schema = await getSampleSchema(getStarWarsGraphQlSchemaV3);
  const type = schema.getType("ReviewInput") as GraphQLInputObjectType;
  expect(generateInputType(type, defaultOptions)).toMatchSnapshot();
});

it("can generate input types for various default values", async function() {
  const schema = await getSampleSchema(
    async () => `
schema {
  query: Query
}
type Query {
  randomTest(value: RandomInput!
             ): String
}

input RandomInput {
  id: ID = "foo"
  i: Int = 15
  f: Float = 2.5
  b: Boolean = true
  l: [Int!] = [5, 4, 3]
  s: String = "evil\\"\\\\\\"string\\\\"
  o: ColorInput = { red: 255, green: 255, blue: 255 }
  n: ColorInput = null
}

input ColorInput {
  red: Int!
  green: Int!
  blue: Int!
}
`
  );
  const type = schema.getType("RandomInput") as GraphQLInputObjectType;
  expect(generateInputType(type, defaultOptions)).toMatchSnapshot();
});

async function getSampleSchema(getSchema: () => Promise<string>) {
  const schema = parseToSchema(await getSchema());
  return schema;
}
