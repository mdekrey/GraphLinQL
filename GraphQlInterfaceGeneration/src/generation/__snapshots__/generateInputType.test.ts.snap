// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can generate input types for various default values 1`] = `
"

public class RandomInput
{
    public string? id { get; set; } = System.Text.Json.JsonSerializer.Deserialize<string?>(\\"\\\\\\"foo\\\\\\"\\");
    public int? i { get; set; } = 15;
    public double? f { get; set; } = 2.5;
    public bool? b { get; set; } = true;
    public IEnumerable<int>? l { get; set; } = System.Text.Json.JsonSerializer.Deserialize<IEnumerable<int>?>(\\"[5,4,3]\\");
    public string? s { get; set; } = \\"evil\\\\\\"\\\\\\\\\\\\\\"string\\\\\\\\\\";
    public ColorInput? o { get; set; } = System.Text.Json.JsonSerializer.Deserialize<ColorInput?>(\\"{\\\\\\"red\\\\\\":255,\\\\\\"green\\\\\\":255,\\\\\\"blue\\\\\\":255}\\");
    public ColorInput? n { get; set; }
}
"
`;

exports[`can generate input types from v3 1`] = `
"
/// <summary>
/// The input object sent when someone is creating a new review
/// </summary>
public class ReviewInput
{
    
      /// <summary>
      /// 0-5 stars
      /// </summary>
      public int stars { get; set; }
    
      /// <summary>
      /// Comment about the movie, optional
      /// </summary>
      public string? commentary { get; set; }
    
      /// <summary>
      /// Favorite color, optional
      /// </summary>
      public ColorInput? favorite_color { get; set; }
}
"
`;
